modules/test-runner/src/swap/swap.test.ts

```diff
@@ -130,4 +130,123 @@ describe("Swaps", () => {
     );
     expect(postSwapFreeBalanceTokenNode).toBeBigNumberEq(swapAmount);
   });
+
+  test("Bot A tries to swap with invalid from token address", async () => {
+    // client deposit and request node collateral
+    await clientA.deposit({ amount: parseEther("0.01").toString(), assetId: AddressZero });
+    await clientA.requestCollateral(tokenAddress);
+
+    const swapRate = await clientA.getLatestSwapRate(AddressZero, tokenAddress);
+    const wrongAddress = "0xdeadbeef";
+    const swapAmount = One;
+    const swapParams: SwapParameters = {
+      amount: swapAmount.toString(),
+      fromAssetId: wrongAddress,
+      swapRate,
+      toAssetId: tokenAddress,
+    };
+    await expect(clientA.swap(swapParams)).rejects.toThrowError(`is not a valid eth address`);
+  });
+
+  test("Bot A tries to swap with invalid to token address", async () => {
+    // client deposit and request node collateral
+    await clientA.deposit({ amount: parseEther("0.01").toString(), assetId: AddressZero });
+    await clientA.requestCollateral(tokenAddress);
+
+    const swapRate = await clientA.getLatestSwapRate(AddressZero, tokenAddress);
+    const wrongAddress = "0xdeadbeef";
+    const swapAmount = One;
+    const swapParams: SwapParameters = {
+      amount: swapAmount.toString(),
+      fromAssetId: AddressZero,
+      swapRate,
+      toAssetId: wrongAddress,
+    };
+    await expect(clientA.swap(swapParams)).rejects.toThrowError(`is not a valid eth address`);
+  });
+
+  test("Bot A tries to swap with insufficient free balance for the user", async () => {
+    // client deposit and request node collateral
+    await clientA.deposit({ amount: parseEther("0.01").toString(), assetId: AddressZero });
+    await clientA.requestCollateral(tokenAddress);
+
+    const swapRate = await clientA.getLatestSwapRate(AddressZero, tokenAddress);
+    const swapAmount = parseEther("0.02");
+    const swapParams: SwapParameters = {
+      amount: swapAmount.toString(),
+      fromAssetId: AddressZero,
+      swapRate,
+      toAssetId: tokenAddress,
+    };
+    await expect(clientA.swap(swapParams)).rejects.toThrowError(`is not less than or equal to`);
+  });
+
+  test("Bot A tries to swap with negative swap rate", async () => {
+    // client deposit and request node collateral
+    await clientA.deposit({ amount: parseEther("0.01").toString(), assetId: AddressZero });
+    await clientA.requestCollateral(tokenAddress);
+
+    const swapRate = await clientA.getLatestSwapRate(AddressZero, tokenAddress);
+    const swapAmount = One;
+    const swapParams: SwapParameters = {
+      amount: swapAmount.toString(),
+      fromAssetId: AddressZero,
+      swapRate: (-swapRate).toString(),
```

Comment:
does this even work?

Reply:
there should be a `BigNumber` function for it that we should definitely use

Reply:
It does work. Swap rate is a string not `BigNumber`
![Screenshot from 2020-01-10 13-04-35](https://user-images.githubusercontent.com/22877754/72140266-cffa5300-33a9-11ea-8014-795dd9b1e366.png)